File: backend/app/api/companies.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session
from app.database import SessionLocal
from app import models
from datetime import datetime, date

router = APIRouter(prefix="/companies", tags=["Companies"])

# -----------------------------------------------------
# 🧠 Database session
# -----------------------------------------------------
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


# -----------------------------------------------------
# 📦 Pydantic Schemas
# -----------------------------------------------------
class CompanyOut(BaseModel):
    id: int
    name: str
    status: Optional[str] = None
    cin_number: Optional[str] = None
    industry_type: Optional[str] = None
    authorized_capital: Optional[float] = None
    paid_up_capital: Optional[float] = None
    address: Optional[str] = None
    registration_date: Optional[date] = None
    director_name: Optional[str] = None
    director_email: Optional[str] = None
    company_secretary: Optional[str] = None
    auditor_name: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        orm_mode = True


class PaginatedCompanies(BaseModel):
    items: List[CompanyOut]
    page: int
    page_size: int
    total: int
    has_next: bool


# -----------------------------------------------------
# 📋 GET /companies — list all companies
# -----------------------------------------------------
@router.get("", response_model=PaginatedCompanies)
def list_companies(
    db: Session = Depends(get_db),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    q: Optional[str] = Query(None, description="Optional search by name or CIN"),
):
    query = db.query(models.Company)

    if q:
        q = q.strip()
        query = query.filter(
            (models.Company.name.ilike(f"%{q}%"))
            | (models.Company.cin_number.ilike(f"%{q}%"))
        )

    total = query.count()
    items = (
        query.order_by(models.Company.id.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    return {
        "items": items,
        "page": page,
        "page_size": page_size,
        "total": total,
        "has_next": (page * page_size) < total,
    }


# -----------------------------------------------------
# 🏗️ POST /companies — create a new company
# -----------------------------------------------------
class CompanyCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    cin_number: Optional[str] = None
    industry_type: Optional[str] = None
    status: Optional[str] = None
    authorized_capital: Optional[float] = None
    paid_up_capital: Optional[float] = None
    address: Optional[str] = None
    registration_date: Optional[date] = None
    director_name: Optional[str] = None
    director_email: Optional[str] = None
    company_secretary: Optional[str] = None
    auditor_name: Optional[str] = None
    notes: Optional[str] = None


@router.post("", response_model=CompanyOut, status_code=201)
def create_company(payload: CompanyCreate, db: Session = Depends(get_db)):
    existing = db.query(models.Company).filter(models.Company.name == payload.name).first()
    if existing:
        raise HTTPException(status_code=400, detail="Company with this name already exists")

    company = models.Company(**payload.dict())
    db.add(company)
    db.commit()
    db.refresh(company)
    return company

File: frontend/src/api/companies.ts
// src/api/companies.ts
import api from "./api";

/** ✅ Company model matching backend schema */
export interface Company {
  id: number;
  name: string;
  cin_number?: string;
  industry_type?: string;
  industry?: string;
  status?: string;
  authorized_capital?: number;
  paid_up_capital?: number;
  address?: string;
  registration_date?: string;
  director_name?: string;
  director_email?: string;
  company_secretary?: string;
  auditor_name?: string;
  notes?: string;
  created_at?: string;
  updated_at?: string;
}

/** ✅ Payload sent when creating/updating a company */
export interface CompanyPayload {
  name: string;
  cin_number?: string | null;
  industry_type?: string | null;
  industry?: string | null;
  status?: string | null;
  authorized_capital?: number | null;
  paid_up_capital?: number | null;
  address?: string | null;
  registration_date?: string | null;
  director_name?: string | null;
  director_email?: string | null;
  company_secretary?: string | null;
  auditor_name?: string | null;
  notes?: string | null;
}

/** ✅ Paginated list of companies */
export interface PaginatedCompanies {
  items: Company[];
  page: number;
  page_size: number;
  total: number;
  has_next: boolean;
}

/** ✅ Fetch all companies (paginated, optional search) */
export async function fetchCompanies(
  page = 1,
  pageSize = 20,
  search = ""
): Promise<PaginatedCompanies> {
  const res = await api.get<PaginatedCompanies>("/companies", {
    params: { page, page_size: pageSize, q: search },
  });
  return res.data;
}

/** ✅ Create a new company (sends structured payload) */
export async function createCompany(payload: CompanyPayload): Promise<Company> {
  const res = await api.post<Company>("/companies", payload);
  return res.data;
}

/** ✅ Update an existing company (sends structured payload) */
export async function updateCompany(
  id: number,
  payload: CompanyPayload
): Promise<Company> {
  const res = await api.put<Company>(`/companies/${id}`, payload);
  return res.data;
}

/** ✅ Delete a company */
export async function deleteCompany(id: number): Promise<void> {
  await api.delete(`/companies/${id}`);
}
